### PR TITLE
bench: Add benchmark for the MLIR parser

### PR DESCRIPTION
**Pull request by math-fehr**
Add a benchmark for the MLIR parser.
This will let us quickly see if we have performance regressions/improvements when parsing MLIR files.
It will also help us see how many MLIR tests we are parsing.

I'm not sure if we want to have this in our CI though, for two reasons:
* I do not trust GitHub machines for benchmarks, and we might have too much noise from there.
* We need to have MLIR installed to first put all files in generic form

### CODE REVIEW

**Comment by math-fehr** [comment]
Also, we currently have this results:

File: bench/parser/bench_parser.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import timeit
+from typing import Iterable
+from xdsl.ir import MLContext
+
+from xdsl.parser import MLIRParser
+
+
+def parse_file(file: str, ctx: MLContext):
+    """
+    Lex the given file
```
Comment (webmiche) [inline bench/parser/bench_parser.py:25]:
```suggestion
    Parse the given file
```

File: bench/parser/bench_parser.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+the generic form.
+Each file is also splitted according to `split-input-file` if it is
+present.
+"""
+
+import subprocess
+import cProfile
+import argparse
+import glob
+import timeit
```
Comment (webmiche) [inline bench/parser/bench_parser.py:16]: Which of these are additional dependencies? Do we need another requirements file?

File: bench/parser/bench_parser.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        print("Parsing file: " + file_name)
+        splitted_contents = split_mlir_file(contents)
+        print(f"File has been split into {len(splitted_contents)} sub-files.")
+
+        # Parse each sub-file separately.
+        for sub_contents in splitted_contents:
+
+            # First, parse the file with MLIR to check that it is valid, and
+            # print it back in generic form.
+            res = subprocess.run([
```
Comment (webmiche) [inline bench/parser/bench_parser.py:61]: Maybe run this with a timeout?

**Review by webmiche — COMMENTED** [review]
I am not quite sure where we want this in the repo. This should certainly not be distributed with Pypi....

**Comment by georgebisbas** [comment]
> Would be cool to have this similar to codecov, where something posts a comment with some performance data that we can then ignore thinking

https://asv.readthedocs.io/en/stable/#

File: bench/parser/bench_parser.py
Comment (georgebisbas) [inline bench/parser/bench_parser.py:1]: Overall I feel this is very hacky, and we should look for other solutions

File: bench/parser/bench_parser.py (hunk 2 above)
Comment (math-fehr) [reply bench/parser/bench_parser.py:16]: Nope, those are in from Python

File: bench/parser/bench_parser.py (hunk 3 above)
Comment (math-fehr) [reply bench/parser/bench_parser.py:61]: Very good point

File: bench/parser/bench_parser.py
Comment (math-fehr) [reply bench/parser/bench_parser.py:1]: Yeah I removed it, so we can just use `python bench_lexer.py` instead.
We could have this in the namespace, but I would like to not pollute it with things most people wouldn't use.

**Comment by math-fehr** [comment]
> > Would be cool to have this similar to codecov, where something posts a comment with some performance data that we can then ignore thinking
> 
> https://asv.readthedocs.io/en/stable/#

If anyone fancy adding this, this could be indeed nice.
However, we should wait that our parser is compatible with MLIR, since right now we are only parsing half of the files. If we then add new features, it will show a slowdown, even if we did nothing wrong.

**Review by superlopuh — APPROVED** [review]
This is a great first step, would be good to use a profiling package eventually, but we'll be able to reuse this code when that happens

**Comment by math-fehr** [comment]
> I am not quite sure where we want this in the repo. This should certainly not be distributed with Pypi....

Can we easily remove it from Pypi? I agree that we shouldn't keep this in PyPi

**Comment by superlopuh** [comment]
I just pip installed in an empty dir and this file is not included. Time to :ship:  :)

**Pull request merged**